hub/core/storage/local.py

```diff
@@ -13,7 +14,137 @@ def __init__(self, root: str):
 
         Args:
             root (str): The root of the provider. All read/write request keys will be appended to root."
+
+        Raises:
+            FileAtRootException: If the root is a file instead of a directory.
+        """
+        if os.path.isfile(root):
+            raise FileAtRootException
+        self.root = root
+
+    def __getitem__(self, path: str):
+        """Gets the object present at the path within the given byte range.
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            my_data = local_provider["abc.txt"]
+
+        Args:
+            path (str): The path relative to the root of the provider.
+
+        Returns:
+            bytes: The bytes of the object present at the path.
+
+        Raises:
+            KeyError: If an object is not found at the path.
+            DirectoryAtPathException: If a directory is found at the path.
+        """
+        try:
+            full_path = self._check_is_file(path)
+            file = open(full_path, "rb")
+            return file.read()
+        except DirectoryAtPathException:
+            raise
+        except Exception:
+            raise KeyError
+
+    def __setitem__(self, path: str, value: bytes):
+        """Sets the object present at the path with the value
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            local_provider["abc.txt"] = b"abcd"
+
+        Args:
+            path (str): the path relative to the root of the provider.
+            value (bytes): the value to be assigned at the path.
+
+        Raises:
+            Exception: If unable to set item due to directory at path or permission or space issues.
+        """
+        try:
+            full_path = self._check_is_file(path)
+            file = open(full_path, "wb")
+            file.write(value)
+        except Exception:
+            raise
+
+    def __delitem__(self, path: str):
+        """Delete the object present at the path.
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            del local_provider["abc.txt"]
+
+        Args:
+            path (str): the path to the object relative to the root of the provider.
+
+        Raises:
+            KeyError: If an object is not found at the path.
+            DirectoryAtPathException: If a directory is found at the path.
+        """
+        try:
+            full_path = self._check_is_file(path)
+            os.remove(full_path)
+        except DirectoryAtPathException:
+            raise
+        except Exception:
+            raise KeyError
```

Comment:
KeyError seems a bit generic. Can we at least tack on an explanatory string here?

It's important to remember that almost all of these exceptions are going to bubble up to the end-user, who has no idea about our internal constructs.